### ISSUE TITLE
Only one of two clients working

### ISSUE THREAD

**Issue by stefaanv** [eclipse/paho.mqtt.m2mqtt#23]
I'm writing a program that translates MQTT messages from one from into another form.
The platform is Windows IOT Core on Raspberry Pi 3, I'm writing an UWP application.
It seems that of the two clients that are connected to the same broker, only one of them is emitting messages.  The `Publish()` action on the second client passes without complaining, but nothing arrives at the broker.  Even though I'm using exactly the same Publish code on both clients.

**Comment by JTrotta** [comment]
Are you using the same clientId? In this case you cannot.
Let's show us some piece  of code.

**Comment by stefaanv** [comment]
Nope, two different classes each use their own client
The code is on https://github.com/stefaanv/ProtocolHubUWP

This is the first class
`    public class MqttFacadeController : IController
    {
        //TODO: referentie naar Gateway uitvlassen
        private List<string> subscriptions;
        private MqttClient mqtt;
        private Action<IMessage> callback;

        public MqttFacadeController(string brokerHostname)
        {
            subscriptions = new List<string>();
            mqtt = new MqttClient(brokerHostname);
            Debug.WriteLine("Subscribed to "+brokerHostname);
            mqtt.Connect(Guid.NewGuid().ToString());
            mqtt.MqttMsgPublishReceived += MqttOnMqttMsgPublishReceived;
        }

        private void MqttOnMqttMsgPublishReceived(object sender, MqttMsgPublishEventArgs e)
        {
            string data = Encoding.ASCII.GetString(e.Message);
            string topic = e.Topic;

            IMessage m = MessageSerializer.DeSerialize(data, topic);
            if (m == null)
                Debug.WriteLine($"MQTT message failed to deserialize : {data.Replace('\n', ' ')}");
            callback(m);
        }

        public void MessageToHub(IMessage message)
        {
            string topic = message.Topic;
            string data = MessageSerializer.Serialize(message);
            byte[] dataAsByteArray = Encoding.UTF8.GetBytes(data);
            Debug.WriteLine($"MQTT OUT : {topic} <> {data}");
            mqtt.Publish(topic, dataAsByteArray, MqttMsgBase.QOS_LEVEL_EXACTLY_ONCE, false);
        }

        public void Subscribe(string[] topics, Action<IMessage> callBack)
        {
            subscriptions.AddRange(topics);
            mqtt.Subscribe(topics,
                new[] {MqttMsgBase.QOS_LEVEL_EXACTLY_ONCE});
            this.callback = callBack;
        }
    }`

This is the second class
`    internal class DomoticzMqttGateway : Gateway<DomoticzMqttFilament, DomoticzMqttFormat>
    {
        private MqttClient mqttClient;

        [JsonProperty("BrokerName", Required = Required.Always)]
        private string brokerName;

        [JsonProperty("Ignore", Required = Required.AllowNull)]
        private IList<DomoticzMqttFilamentKey> ignoreList;

        [JsonProperty("DomoticzOutTopic", DefaultValueHandling = DefaultValueHandling.Populate)]
        [DefaultValue("domoticz/out")]
        private string domoticzOutTopic;

        [JsonProperty("DomoticzInTopic", DefaultValueHandling = DefaultValueHandling.Populate)]
        [DefaultValue("domoticz/in")]
        private string domoticzInTopic;

        [DefaultValue(1000)]
        [JsonProperty("EchoCancellationTime", DefaultValueHandling = DefaultValueHandling.Populate)]
        private int echoCancellationTimeMs;

        private TimeSpan echoCancellationTimeSpan;  //TODO: moet nog in configbestand komen !!!
        private DomoticzMqttPacketTranslator translator;

        #region Constructor
        public DomoticzMqttGateway(/*ILog log, IController controller*/)
               : base(/*log, controller*/)
        {
            translator = new DomoticzMqttPacketTranslator();
        }
        #endregion

        #region Abstract method implementations - template method pattern
        protected override bool TrySetupHardwarePrimitive()
        {
            echoCancellationTimeSpan = new TimeSpan(echoCancellationTimeMs * 10000);
            try
            {
                Debug.WriteLine("Program version 2");
                Debug.WriteLine("Subscribing DomoticzMqttGateway to " + brokerName);
                mqttClient = new MqttClient(brokerName);
                mqttClient.ConnectionClosed += MqttClientConnectionClosed;
                mqttClient.MqttMsgPublishReceived += MqttClientMqttMsgPublishReceived;
                return true;
            }
            catch (System.Net.Sockets.SocketException)
            {
                mqttClient = null;
            }
            return false;
        }

        protected override bool TryConnectPrimitive()
        {
            try
            {
                mqttClient.Connect(Guid.NewGuid().ToString());
                mqttClient.Subscribe(new[] { domoticzOutTopic }, new[] { MqttMsgBase.QOS_LEVEL_EXACTLY_ONCE });
            }
            catch (Exception)
            {
                Debug.WriteLine("Unable to connect to broker, trying again ...");
                return false;
            }

            if (!mqttClient.IsConnected) return false;
            State = GatewayState.Connected;
            return mqttClient.IsConnected;
        }

        protected override void DisconnectPrimitive()
        {
            mqttClient.Unsubscribe(new[] { domoticzOutTopic });
            mqttClient.Disconnect();
        }

        protected override void CleanupHardwarePrimitive()
        {
            mqttClient.ConnectionClosed -= MqttClientConnectionClosed;
            mqttClient.MqttMsgPublishReceived -= MqttClientMqttMsgPublishReceived;
            mqttClient = null;
        }

        protected override IMessage TranslatePrimitive(DomoticzMqttFormat m, DomoticzMqttFilament filament)
        {
            var convertedMessage = translator.ConvertToInternal(m, filament, this);
            bool cancelEcho = filament != null 
                && (DateTimeOffset.Now < filament.LastReceivedTime.Add(echoCancellationTimeSpan));
            if (convertedMessage == null) return null;

            if (cancelEcho)
            {
                filament.LastReceivedTime = DateTimeOffset.Now.Subtract(echoCancellationTimeSpan); // Only cancel first the (re-)occurrence
                return null;
            }

            return convertedMessage;
        }
        #endregion

        private void MqttClientMqttMsgPublishReceived(object sender, MqttMsgPublishEventArgs e)
        {
            string data = Encoding.ASCII.GetString(e.Message);
            string topic = e.Topic;

            if (topic != domoticzOutTopic) return;

            var m = JsonConvert.DeserializeObject<DomoticzMqttFormat>(data);
                int idx = m.idx;
                var exKey = new DomoticzMqttFilamentKey(idx);
                if (ignoreList.Contains(exKey))
                {
                    Debug.WriteLine($"Message IGNORED Idx={idx} {data.CleanForDisplay()}");
                    return;
                }
                MessageToHub(m, exKey,
                    $"DomoticzMqtt filament with idx={idx} not found in {Filaments.Count} filaments");
        }

        private void MqttClientConnectionClosed(object sender, EventArgs e)
        {
            State = GatewayState.PermanentError;
            ReconnectAsync().Wait();
        }

        protected override void MessageFromHubPrimitive(IMessage message, IFilament filament)
        {
            if (State != GatewayState.Connected) return;
            var connection = (DomoticzMqttFilament)filament;
            string convertedMessage = translator.ConvertFromInternal(message, connection);
            connection.LastReceivedTime = DateTimeOffset.Now;
            var msgBytes = Encoding.ASCII.GetBytes(convertedMessage);
            mqttClient.Publish(domoticzInTopic, msgBytes, MqttMsgBase.QOS_LEVEL_EXACTLY_ONCE, false);
        }
    }
`

**Comment by JTrotta** [comment]
Sorry but cannot find the problem. Are you sure the second client is really connected? Check on the Broker if the second client is really connected. I'm using tents of clients with no problem.

**Comment by stefaanv** [comment]
Yes I know.  I wrote a super simple program on the same platform that uses 2 client that alternatively publish a message and it works fine.  I keep researching the problem.

**Comment by stefaanv** [comment]
Still no luck, but I see that the InFlightQueue builds up.
Is it possible that the messages somehow are stored in the system but not
really sent ?
I do send test message now immediately after creating the clients, these
test messages do get sent.

2016-12-13 14:34 GMT+01:00 JTrotta <notifications@github.com>:

**Comment by stefaanv** [comment]
Really strange.  Test messages immediately after connecting do get through.
However, I see that the inflightQueue is filling up with the messages that don't get sent.
Pointed the second client to a different broker ... same result.
Checked traffic coming into the second broker with Wireshark, The test messages immediate after connecting show up in Wireshark.  They also het configmed by the broker.  Succeeding message don't show up on wireshark.  Seems there is not even an attempt to send these.

**Comment by stefaanv** [comment]
Nailed the problem down.
This problem only occurs when I as the MqttClient to subscribe to many topics in 1 Subscribe operation by passing an array of strings with more then one element.
When I call the Subscribe method as many times as I have topics to subscribe to, each time with an array of 1, then I don't have the problem.
